refl1d/magnetism.py

```diff
@@ -437,3 +437,150 @@ def __str__(self):
 
     def __repr__(self):
         return "FreeMagnetism"
+
+
+@dataclass(init=False)
+class FreeMagnetismInterface(BaseMagnetism):
+    r"""
+    Spline change in magnetism throughout layer.
+
+    Defines monotonic splines for rhoM and thetaM with shared knot positions.
+
+    *dz* is the relative $z$ step between the knots, with position $z_k$
+    defined by $z_k = w \sum_{i=0}^k \delta z_i / \sum_{i=0}^n \delta z_i$,
+    where $n$ is the number of intervals. The resulting $z$ must be monotonic,
+    with $\delta z_i \ge 0$ for all intervals.
+
+    *drhoM* gives the relative $\rho_M$ step between knots. Unlike
+    $\rho_{Mk} = \sum_{i=0}^k \delta \rho_{Mi} / \sum_{i=0}^n \delta \delta \rho_{Mi}$.
+
+    *dthetaM* gives the magnetic angle for each knot.
+
+    *name* is the base name for the various layer parameters.
+
+    *dead_above* and *dead_below* define magnetically dead layers at the
+    nuclear boundaries.  These can be negative if magnetism extends beyond
+    the nuclear boundary.
+
+    *interface_above* and *interface_below* define the magnetic interface
+    at the boundaries, if it is different from the nuclear interface.
+
+    *mbelow* and *mabove* are the rhoM parameter values of
+    the above and below layers respectively. Do not specify if
+    layers either side are not magnetic.
+
+    *tbelow* and *tabove* are the thetaM parameter values of
+    the above and below layers respectively. Do not specify if
+    layers either side are not magnetic.
+    """
+
+    name: Optional[str]
+    mbelow: Optional[Union[Parameter, float]]
```

Comment:
@bmaranville I am not sure if this is the current style that you are using for typing. I.e. would you prefer to remove `Any` and `Optional`?

Reply:
It's not really a question of style - the dataclass attribute types should be exactly what will be stored in those attributes after init + post_init.  There can be different types for the arguments of the init function, if they are going to be coerced before being stored in the attribute.  For example, an argument will often be of type `Union[float, Parameter]` but then we store it as `Parameter.default(value)` which coerces it to a Parameter, so that should be the type of the attribute (not the Union)

Reply:
The typing of the arg is not the same as the typing of the attribute - `tbelow` and `tabove` are both initialized as `Parameter` objects during init, even if they are `None` as an argument.  The type of the attributes should be `Parameter`